### DISCUSSION TITLE
Crop to bounding box of the mask

### DISCUSSION THREAD

**Discussion by mattwarkentin** [TorchIO-project/torchio#677] [Q&A]
Hi @fepegar,

Is there a way to crop to the bounding box of the non-zero values in a mask?

Such as with data like:
```py
import torchio as tio

sub = tio.Subject(
  image = tio.ScalarImage(...),
  mask = tio.LabelMap(...)
)

# Functionality like
tio.Crop('mask')(sub)
```

**Comment by romainVala** [comment]
Hello
No currently, only the `CropOrPad` transform can take a `mask_name` argument in addition to `target_shape` 
The center of mass of the mask is then used to choose the center of the FOV (field of view) of the corped or paded volume
I thing it should do the job, you just have to guess the final target shape you wish 

If you implement, it , I would go with CropOrPad, but the risk is then if mask are changing from subject to subject, you may end up with different crop between subject and thus a different shape, which is not a good idea

I hope it helps

**Comment by fepegar** [comment]
This seems to be related to https://github.com/fepegar/torchio/issues/675

@mattwarkentin, can you tell us a bit more about what you're planning to do?

**Comment by mattwarkentin** [comment]
> you just have to guess the final target shape you wish

Indeed this is what I am trying to avoid. By choosing a fixed size, you will end up with more or less (perhaps undesired) volume surrounding the ROI.

> but the risk is then if mask are changing from subject to subject, you may end up with different crop between subject and thus a different shape

This is true, but I think when paired with the recently added `tio.Resize` transform it can work well.

> can you tell us a bit more about what you're planning to do?

@fepegar I think there are use-cases where one might want to crop an ROI to "fill" the FOV. In other words, even if the ROI is a different size across subjects, the cropped volume will have the ROI centred and generally filling the cropped volume space. In the example of #675, one may wish to crop down to the bounding box of the lung mask, and then resize so that the output volumes for all subjects are equivalent in shape.

For example,
```py
sub = tio.Subject(
    image = tio.ScalarImage(...),
    mask = tio.LabelMap(...)
)

trans = tio.Compose([
    tio.CropToMask('mask'),
    tio.Resize((100, 100, 100))
])

sub_t = trans(sub)
```

In this case, you will always get a consistent volume size that contains the entire ROI for every subject; and the ROI will generally fill the viewport. Hopefully that makes sense. Please let me know if you think this is not sensible.

**Reply by mattwarkentin** [reply]
FWIW, I think [`monai.transforms.CropForeground`](https://docs.monai.io/en/latest/transforms.html#cropforeground) is similar in nature to what this feature would entail.

**Reply by fepegar** [reply]
I think it's almost never a good idea to use `Resize`, you might very easily get unexpected results, `Resample` is typically a better choice. But I guess that's another topic.

So if I understand correctly, you would like the FOV of images in the subject to match the bounding box of one of the label maps in the subject, right? This might actually pad that image, so I think it should be either a new feature of `CropOrPad` or a new transform that calls it.

**Reply by mattwarkentin** [reply]
> I think it's almost never a good idea to use Resize, you might very easily get unexpected results

I almost always use `Resample`, but could you elaborate on the potential risk of using `Resize`?

> So if I understand correctly, you would like the FOV of images in the subject to match the bounding box of one of the label maps in the subject, right? 

Yes, this is correct.

> This might actually pad that image, so I think it should be either a new feature of CropOrPad or a new transform that calls it.

I don't think there could be any padding unless a `margin` argument is support, right? Well I guess I am assuming that the image and label map are the same size and the ROI is a subvolume.

**Reply by fepegar** [reply]
> I almost always use Resample, but could you elaborate on the potential risk of using Resize?

I guess the main problem is that when you resize a _medical_ image, you might be deforming it in the physical space. This is why I added it only recently, and I'm not really happy about it because beginners in medical imaging with experience in computer vision will tend to use it instead of `Resample` (which does _not_ change the aspect ratio of the objects in the physical space. I hope that helps a bit, there is some more info in the PR: https://github.com/fepegar/torchio/pull/642

> Yes, this is correct.

Ok, thanks. I'll try to work on this soon.

> I don't think there could be any padding unless a margin argument is support, right? Well I guess I am assuming that the image and label map are the same size and the ROI is a subvolume.

Exactly, I think we have the tools to make it generic enough to assume that the ROI might actually be larger than the FOV of the scalar image (as `CropOrPad` already does this, I think).

**Reply by mattwarkentin** [reply]
Thanks for the explanation.

> you would like the FOV of images in the subject to match the bounding box of one of the label maps in the subject, right?

A label map may include multiple integer labels for several ROI, and so the bounding box may only be for a single label, or a collection of labels. Similar to `tio.Mask(labels=[...])`. So that the bounding box is the box that contains ALL labels in `labels`. Do you agree?

**Reply by fepegar** [reply]
Hmm. Are you requesting support to specify the labels in the label map to compute the ROI?

**Reply by mattwarkentin** [reply]
I suppose. If we are cropping down to a bounding box, the bounding box could be for a single label in the label map or for a set of labels (similar to the `select_fn` in `CropForeground`).

**Reply by fepegar** [reply]
Alright! How's this?

```python
import torchio as tio
cp = tio.CropOrPad(mask_name='brain')
colin = tio.datasets.Colin27()
colin_tight_brain = cp(colin)
colin_tight_brain.plot()
```

![brain](https://user-images.githubusercontent.com/12688084/135651652-d7f41337-7a2e-430e-ad48-14330d2f7c18.png)

```python
import torchio as tio
cp = tio.CropOrPad(mask_name='cls', labels=[3])
colin = tio.datasets.Colin27(2008)
colin_tight_wm = cp(colin)
colin_tight_wm.plot()
```
![wm](https://user-images.githubusercontent.com/12688084/135651680-f2eac34c-eb72-4c78-af92-9f7936e98d27.png)

**Reply by mattwarkentin** [reply]
Wow, that was fast. LGTM!

**Reply by fepegar** [reply]
Kudos to @GReguig, @efirdc and @Svdvoort, who contributed the modular features in `CropOrPad`, the labels transforms and `Mask` that makes these changes easy.

I will open a PR in a second.
